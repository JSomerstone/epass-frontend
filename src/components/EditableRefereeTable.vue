<template>
  <b-table :data="referees" class="no-margin-on-control" hoverable>
    <template slot-scope="props">
        <b-table-column field="name" label="Name" width="190">
            {{ props.row.firstName }}  {{ props.row.lastName }}
        </b-table-column>
        <b-table-column field="games" label="Games as referee" centered>
          <b-numberinput 
            controls-position="compact" 
            v-model="props.row.games"
            @input="update(props.row)"
            v-if="isEditable(props.row)"
          />
          <span v-else>{{ props.row.games }}</span>
        </b-table-column>
        <b-table-column field="tenSeconds" label="Games as 10s / official" centered>
          <b-numberinput 
            controls-position="compact" 
            v-model="props.row.tenSeconds"
            @input="update(props.row)"
            v-if="isEditable(props.row)"
          />
          <span v-else>{{ props.row.tenSeconds }}</span>
        </b-table-column>
        <b-table-column field="id" v-if="editable">
          <b-tooltip 
            label="Remove referee from tournament"
            position="is-bottom"
            type="is-danger"
            v-if="isEditable(props.row)"
          >
            <b-button 
              @click="removeRow(props.row)" 
              icon-left="close"
              type="is-danger"
              outlined
            />
          </b-tooltip>
        </b-table-column>
      </template>
  </b-table>
</template>
<style>
.no-margin-on-control .field p.control {
  margin-bottom: 0px;
}
</style>
<script>
export default {
  data(){
    return {
      referees: this.value
    };
  },
  props: {
    value: {
      type: Array,
      required: true,
    },
    editable: {
      type: Boolean,
      default: true,
    },
    editableItem: {
      type: String,
      default: "all"
    }
  },
  methods: {
    isEditable: function(row) {
      return this.editable && (this.editableItem == "all" || this.editableItem == row.id);
    },
    update({ id, games = 0, tenSeconds = 0 }) {
      const indx = this.referees.find( r => r.id == id);
      this.referees[indx] = {
        ...this.referees[indx],
        games,
        tenSeconds
      };
      
      this.$emit('input', this.referees);
    }
  }
}
</script>