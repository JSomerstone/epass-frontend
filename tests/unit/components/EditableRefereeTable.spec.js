import { mount } from "@vue/test-utils";
import localVue from "../localVue";
import { BTable }from 'buefy'
import EditableRefereeTable from "@/components/EditableRefereeTable";

describe("EditableRefereeTable", () => {

  let referees = [
    { id: "a", firstName: 'John', lastName: "Connor", games: 1, tenSeconds: 2 },
    { id: "b", firstName: 'Sarah', lastName: "Connor", games: 3, tenSeconds: 4 },
  ];

  it("Renders list of referees when non-editable", () => {
    const wrapper = mount(EditableRefereeTable, {
      localVue,
      propsData: {
        value: referees,
        editable: false,
        editableItem: "none",
      }
    });
    expect(wrapper).toMatchSnapshot();
  });

  it("Renders correct referee as editable", () => {
    const wrapper = mount(EditableRefereeTable, {
      localVue,
      propsData: {
        value: referees,
        editable: true,
        editableItem: "b",
      }
    });
    expect(wrapper).toMatchSnapshot();
  });

  it("Renders all referees as editable", () => {
    const wrapper = mount(EditableRefereeTable, {
      localVue,
      propsData: {
        value: referees,
        editable: true,
        editableItem: "all",
      }
    });
    expect(wrapper).toMatchSnapshot();
  });


  it("Is able to remove an item from it's list", async () => {
    const wrapper = mount(EditableRefereeTable, {
      localVue,
      propsData: {
        value: referees,
        editable: true,
        editableItem: "b",
      }
    });
    const button = wrapper.find('button.is-danger')
    await button.trigger('click')
    expect(wrapper).toMatchSnapshot();
  });
});