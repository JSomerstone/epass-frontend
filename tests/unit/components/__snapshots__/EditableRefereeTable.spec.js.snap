// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableRefereeTable Is able to remove an item from it's list 1`] = `
<div class="b-table no-margin-on-control referee-table">
  <!---->
  <!---->
  <div class="table-wrapper has-mobile-cards">
    <table class="table is-hoverable">
      <thead>
        <tr>
          <!---->
          <!---->
          <th class="">
            <div class="th-wrap">Name
              <!---->
            </div>
          </th>
          <th class="" style="width: 140px;">
            <div class="th-wrap is-centered">Games as referee
              <!---->
            </div>
          </th>
          <th class="" style="width: 140px;">
            <div class="th-wrap is-centered">Games as 10s / official
              <!---->
            </div>
          </th>
          <th class="">
            <div class="th-wrap">
              <!---->
            </div>
          </th>
          <!---->
        </tr>
        <!---->
        <!---->
      </thead>
      <tbody>
        <tr draggable="false" class="">
          <!---->
          <!---->
          <td data-label="Name" class="">
            John Connor
          </td>
          <td data-label="Games as referee" class="has-text-centered"><span>1</span></td>
          <td data-label="Games as 10s / official" class="has-text-centered"><span>2</span></td>
          <td class=""></td>
          <!---->
        </tr>
        <!---->
        <!---->
      </tbody>
      <!---->
    </table>
  </div>
  <!---->
</div>
`;

exports[`EditableRefereeTable Renders all referees as editable 1`] = `
<div class="b-table no-margin-on-control referee-table">
  <!---->
  <!---->
  <div class="table-wrapper has-mobile-cards">
    <table class="table is-hoverable">
      <!---->
      <tbody>
        <tr draggable="false" class="">
          <!---->
          <!---->
          <td data-label="Name" class="">
            John Connor
          </td>
          <td data-label="Games as referee" class="has-text-centered">
            <div class="b-numberinput field has-addons">
              <p class="control"><button type="button" class="button is-primary"><span class="icon"><i class="mdi mdi-minus mdi-24px"></i></span></button></p>
              <div class="control is-clearfix"><input type="number" autocomplete="on" step="1" class="input">
                <!---->
                <!---->
                <!---->
              </div>
              <p class="control"><button type="button" class="button is-primary"><span class="icon"><i class="mdi mdi-plus mdi-24px"></i></span></button></p>
            </div>
          </td>
          <td data-label="Games as 10s / official" class="has-text-centered">
            <div class="b-numberinput field has-addons">
              <p class="control"><button type="button" class="button is-primary"><span class="icon"><i class="mdi mdi-minus mdi-24px"></i></span></button></p>
              <div class="control is-clearfix"><input type="number" autocomplete="on" step="1" class="input">
                <!---->
                <!---->
                <!---->
              </div>
              <p class="control"><button type="button" class="button is-primary"><span class="icon"><i class="mdi mdi-plus mdi-24px"></i></span></button></p>
            </div>
          </td>
          <td class=""><span data-label="Remove referee from tournament" class="is-danger is-bottom is-medium b-tooltip" style="transition-delay: 0ms;"><button type="button" class="button is-danger is-outlined"><span class="icon is-small"><i class="mdi mdi-close"></i></span>
            <!---->
            <!----></button></span></td>
          <!---->
        </tr>
        <!---->
        <!---->
        <tr draggable="false" class="">
          <!---->
          <!---->
          <td data-label="Name" class="">
            Sarah Connor
          </td>
          <td data-label="Games as referee" class="has-text-centered">
            <div class="b-numberinput field has-addons">
              <p class="control"><button type="button" class="button is-primary"><span class="icon"><i class="mdi mdi-minus mdi-24px"></i></span></button></p>
              <div class="control is-clearfix"><input type="number" autocomplete="on" step="1" class="input">
                <!---->
                <!---->
                <!---->
              </div>
              <p class="control"><button type="button" class="button is-primary"><span class="icon"><i class="mdi mdi-plus mdi-24px"></i></span></button></p>
            </div>
          </td>
          <td data-label="Games as 10s / official" class="has-text-centered">
            <div class="b-numberinput field has-addons">
              <p class="control"><button type="button" class="button is-primary"><span class="icon"><i class="mdi mdi-minus mdi-24px"></i></span></button></p>
              <div class="control is-clearfix"><input type="number" autocomplete="on" step="1" class="input">
                <!---->
                <!---->
                <!---->
              </div>
              <p class="control"><button type="button" class="button is-primary"><span class="icon"><i class="mdi mdi-plus mdi-24px"></i></span></button></p>
            </div>
          </td>
          <td class=""><span data-label="Remove referee from tournament" class="is-danger is-bottom is-medium b-tooltip" style="transition-delay: 0ms;"><button type="button" class="button is-danger is-outlined"><span class="icon is-small"><i class="mdi mdi-close"></i></span>
            <!---->
            <!----></button></span></td>
          <!---->
        </tr>
        <!---->
        <!---->
      </tbody>
      <!---->
    </table>
  </div>
  <!---->
</div>
`;

exports[`EditableRefereeTable Renders correct referee as editable 1`] = `
<div class="b-table no-margin-on-control referee-table">
  <!---->
  <!---->
  <div class="table-wrapper has-mobile-cards">
    <table class="table is-hoverable">
      <!---->
      <tbody>
        <tr draggable="false" class="">
          <!---->
          <!---->
          <td data-label="Name" class="">
            John Connor
          </td>
          <td data-label="Games as referee" class="has-text-centered"><span>1</span></td>
          <td data-label="Games as 10s / official" class="has-text-centered"><span>2</span></td>
          <td class=""></td>
          <!---->
        </tr>
        <!---->
        <!---->
        <tr draggable="false" class="">
          <!---->
          <!---->
          <td data-label="Name" class="">
            Sarah Connor
          </td>
          <td data-label="Games as referee" class="has-text-centered">
            <div class="b-numberinput field has-addons">
              <p class="control"><button type="button" class="button is-primary"><span class="icon"><i class="mdi mdi-minus mdi-24px"></i></span></button></p>
              <div class="control is-clearfix"><input type="number" autocomplete="on" step="1" class="input">
                <!---->
                <!---->
                <!---->
              </div>
              <p class="control"><button type="button" class="button is-primary"><span class="icon"><i class="mdi mdi-plus mdi-24px"></i></span></button></p>
            </div>
          </td>
          <td data-label="Games as 10s / official" class="has-text-centered">
            <div class="b-numberinput field has-addons">
              <p class="control"><button type="button" class="button is-primary"><span class="icon"><i class="mdi mdi-minus mdi-24px"></i></span></button></p>
              <div class="control is-clearfix"><input type="number" autocomplete="on" step="1" class="input">
                <!---->
                <!---->
                <!---->
              </div>
              <p class="control"><button type="button" class="button is-primary"><span class="icon"><i class="mdi mdi-plus mdi-24px"></i></span></button></p>
            </div>
          </td>
          <td class=""><span data-label="Remove referee from tournament" class="is-danger is-bottom is-medium b-tooltip" style="transition-delay: 0ms;"><button type="button" class="button is-danger is-outlined"><span class="icon is-small"><i class="mdi mdi-close"></i></span>
            <!---->
            <!----></button></span></td>
          <!---->
        </tr>
        <!---->
        <!---->
      </tbody>
      <!---->
    </table>
  </div>
  <!---->
</div>
`;

exports[`EditableRefereeTable Renders list of referees when non-editable 1`] = `
<div class="b-table no-margin-on-control referee-table">
  <!---->
  <!---->
  <div class="table-wrapper has-mobile-cards">
    <table class="table is-hoverable">
      <!---->
      <tbody>
        <tr draggable="false" class="">
          <!---->
          <!---->
          <td data-label="Name" class="">
            John Connor
          </td>
          <td data-label="Games as referee" class="has-text-centered"><span>1</span></td>
          <td data-label="Games as 10s / official" class="has-text-centered"><span>2</span></td>
          <!---->
          <!---->
        </tr>
        <!---->
        <!---->
        <tr draggable="false" class="">
          <!---->
          <!---->
          <td data-label="Name" class="">
            Sarah Connor
          </td>
          <td data-label="Games as referee" class="has-text-centered"><span>3</span></td>
          <td data-label="Games as 10s / official" class="has-text-centered"><span>4</span></td>
          <!---->
          <!---->
        </tr>
        <!---->
        <!---->
      </tbody>
      <!---->
    </table>
  </div>
  <!---->
</div>
`;
